farm/contract/lib.rs

```diff
@@ -293,25 +295,28 @@ mod farm {
         ) -> Result<(), FarmError> {
             ensure!(self.env().caller() == self.owner, FarmError::CallerNotOwner);
             self.update()?;
-            ensure!(!self.is_active(), FarmError::FarmAlreadyRunning);
+            ensure!(!self.is_active, FarmError::FarmIsRunning);
             self.farm_reward_rates = self.assert_start_params(start, end, rewards.clone())?;
             self.start = start;
             self.end = end;
+            self.is_active = true;
             Ok(())
         }
 
         #[ink(message)]
         fn owner_stop_farm(&mut self) -> Result<(), FarmError> {
             ensure!(self.env().caller() == self.owner, FarmError::CallerNotOwner);
+            ensure!(self.is_active, FarmError::FarmAlreadyStopped);
             self.update()?;
             self.end = self.env().block_timestamp();
```

Comment:
I think we should not set `self.end` anymore if we have `self.is_active`. Instead we can just add a simple check in `update` and that will cover it (something like "if not active, then update does nothing except updating the timestamp"). The downside of setting `self.end` like this is that the state then looks a little weird. I mean even if the owner stops the farm after it has naturally ended, this shifts the `end` to a different timestamp. And if the stop happens before the farm ended it's even weirder, because then `self.start > self.end`.